telegram/ext/contexttypes.py

```diff
@@ -0,0 +1,204 @@
+        ...
+
+    @overload
+    def __init__(self: 'ContextTypes[CallbackContext, Dict, CD, Dict]', chat_data: Type[CD]):
+        ...
+
+    @overload
+    def __init__(self: 'ContextTypes[CallbackContext, Dict, Dict, BD]', user_data: Type[BD]):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, UD, Dict, Dict]', context: Type[CCT], bot_data: Type[UD]
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, Dict, CD, Dict]', context: Type[CCT], chat_data: Type[CD]
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, Dict, Dict, BD]', context: Type[CCT], user_data: Type[BD]
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CallbackContext, UD, CD, Dict]',
+        bot_data: Type[UD],
+        chat_data: Type[CD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CallbackContext, UD, Dict, BD]',
+        bot_data: Type[UD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CallbackContext, Dict, CD, BD]',
+        chat_data: Type[CD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, UD, CD, Dict]',
+        context: Type[CCT],
+        bot_data: Type[UD],
+        chat_data: Type[CD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, UD, Dict, BD]',
+        context: Type[CCT],
+        bot_data: Type[UD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, Dict, CD, BD]',
+        context: Type[CCT],
+        chat_data: Type[CD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CallbackContext, UD, CD, BD]',
+        bot_data: Type[UD],
+        chat_data: Type[CD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: 'ContextTypes[CCT, UD, CD, BD]',
+        context: Type[CCT],
+        bot_data: Type[UD],
+        chat_data: Type[CD],
+        user_data: Type[BD],
+    ):
+        ...
+
+    def __init__(  # type: ignore[no-untyped-def]
+        self,
+        context=CallbackContext,
+        bot_data=dict,
+        chat_data=dict,
+        user_data=dict,
+    ):
+        if not issubclass(context, CallbackContext):
+            raise ValueError('context must be a subclass of CallbackContext.')
+
+        self._context = context
+        self._bot_data = bot_data
+        self._chat_data = chat_data
+        self._user_data = user_data
+
+    @property
+    def context(self) -> Type[CCT]:
+        return self._context
+
+    @context.setter
+    def context(self, value: Any) -> NoReturn:
+        raise AttributeError("You can not assign a new value to ContextTypes attributes.")
+
```

Comment:
Mh, our error message is slightly more informative than `AttributeError: can't set attribute`. I copied that behavior from `Defaults`. Back then, Pieter wanted that , IIRC. But I'd also be fine with dropping those for sake of brevity

Reply:
imo it's better to drop since: 
1. IDE's warn about this if you try to set a read only attribute
2. the traceback tells you can't set an attribute to that anyway
3. It's extra code to maintain :D

Reply:
fair enough. will remove :)